<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>deepG is a R library that contains a collection of scripts required for model generation and model analysis • deepG</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="deepG is a R library that contains a collection of scripts required for model generation and model analysis">
<meta property="og:description" content="deepG is a R library that contains a collection of scripts required for model generation and model analysis.">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">deepG</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="contents col-md-9">


<div id="overview" class="section level2">
<h2 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h2>
<p>deepG is a package for generating LSTM models from genomic text and provides scripts for various common tasks such as the extraction of cell responses. It also comes with example datasets of genomic and human readable languages for testing.</p>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<div id="dependencies" class="section level3">
<h3 class="hasAnchor">
<a href="#dependencies" class="anchor"></a>Dependencies</h3>
<p>For model saving, this package requires the <code>hdf5r</code> library which is based on hdf5 binaries. You can install it from source using the following commands.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="fu">wget</span> https://support.hdfgroup.org/ftp/HDF5/releases/hdf5-1.10/hdf5-1.10.5/src/hdf5-1.10.5.tar.gz</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="fu">tar</span> xvzf hdf5-1.10.5.tar.gz</a>
<a class="sourceLine" id="cb1-3" data-line-number="3"><span class="bu">cd</span> hdf5-1.10.5</a>
<a class="sourceLine" id="cb1-4" data-line-number="4"><span class="ex">./configure</span> --prefix=/usr --enable-fortran --enable-cxx</a>
<a class="sourceLine" id="cb1-5" data-line-number="5"><span class="fu">make</span></a>
<a class="sourceLine" id="cb1-6" data-line-number="6"><span class="fu">make</span> check</a>
<a class="sourceLine" id="cb1-7" data-line-number="7"><span class="fu">sudo</span> make install</a></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/install.packages">install.packages</a></span>(<span class="st">"hdf5r"</span>, <span class="dt">configure.args=</span><span class="st">"--with-hdf5=/usr/bin/h5cc"</span>)</a></code></pre></div>
</div>
</div>
<div id="library" class="section level2">
<h2 class="hasAnchor">
<a href="#library" class="anchor"></a>Library</h2>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/install.packages">install.packages</a></span>(<span class="st">"devtools"</span>)</a>
<a class="sourceLine" id="cb3-2" data-line-number="2">devtools<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/devtools/topics/reexports">install_github</a></span>(<span class="st">"hiddengenome/deepG"</span>)</a></code></pre></div>
<div id="enable-gpu-support" class="section level3">
<h3 class="hasAnchor">
<a href="#enable-gpu-support" class="anchor"></a>Enable GPU support</h3>
<p>On default, Keras will be installed without GPU support. To support GPUs reinstall Keras via</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/install.packages">install.packages</a></span>(<span class="st">"keras"</span>)</a>
<a class="sourceLine" id="cb4-2" data-line-number="2">keras<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/keras/topics/install_keras">install_keras</a></span>(<span class="dt">tensorflow =</span> <span class="st">"gpu"</span>)</a></code></pre></div>
<p>on AWS instances the GPU version can be installed via <code>conda create -y --name r-tensorflow tensorflow-gpu python=3.6.8</code></p>
</div>
</div>
<div id="usage" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#usage" class="anchor"></a>Usage</h1></div>
<div id="generating-of-genomenet" class="section level2">
<h2 class="hasAnchor">
<a href="#generating-of-genomenet" class="anchor"></a>generating of GenomeNet</h2>
<div id="generate-a-genomic-language-model-from-a-collection-of-fasta-files" class="section level3">
<h3 class="hasAnchor">
<a href="#generate-a-genomic-language-model-from-a-collection-of-fasta-files" class="anchor"></a>Generate a genomic language model from a collection of FASTA files</h3>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1">history &lt;-<span class="st"> </span><span class="kw"><a href="reference/trainNetwork.html">trainNetwork</a></span>(<span class="dt">path =</span> <span class="st">"input_dir"</span>)</a></code></pre></div>
<p>When you have multiple GPUs available you can run</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1">history &lt;-<span class="st"> </span><span class="kw"><a href="reference/trainNetwork.html">trainNetwork</a></span>(<span class="dt">path =</span> <span class="st">"input_dir"</span>, <span class="dt">cudnn =</span> T, <span class="dt">multiple_gpu =</span> T, <span class="dt">gpu_num =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">8</span>, <span class="dt">run_name=</span> <span class="st">"GenomeNet"</span>, <span class="dt">epochs =</span> <span class="dv">100</span>, <span class="dt">steps_per_epoch =</span> <span class="dv">10000</span>)</a></code></pre></div>
<p>See the <code><a href="reference/trainNetwork.html">?trainNetwork</a></code> for further information how to setup file names and network size.</p>
</div>
<div id="generate-a-genomic-language-model-using-data-held-in-the-ram" class="section level3">
<h3 class="hasAnchor">
<a href="#generate-a-genomic-language-model-using-data-held-in-the-ram" class="anchor"></a>Generate a genomic language model using data held in the RAM</h3>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1"><span class="co"># load example dataset</span></a>
<a class="sourceLine" id="cb7-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/data">data</a></span>(crispr_sample)</a>
<a class="sourceLine" id="cb7-3" data-line-number="3"></a>
<a class="sourceLine" id="cb7-4" data-line-number="4"><span class="co"># generate one-hot encoding of 30-nucleotide sized chunks</span></a>
<a class="sourceLine" id="cb7-5" data-line-number="5">crispr_preprocessed &lt;-<span class="st"> </span><span class="kw"><a href="reference/preprocessSemiRedundant.html">preprocessSemiRedundant</a></span>(crispr_sample)</a>
<a class="sourceLine" id="cb7-6" data-line-number="6">history &lt;-<span class="st"> </span><span class="kw"><a href="reference/trainNetwork.html">trainNetwork</a></span>(<span class="dt">dataset =</span> crispr_preprocessed)</a></code></pre></div>
</div>
</div>
<div id="train-a-deep-network-for-multiclass-prediction" class="section level2">
<h2 class="hasAnchor">
<a href="#train-a-deep-network-for-multiclass-prediction" class="anchor"></a>Train a deep network for multiclass prediction</h2>
<p>here the FASTA files are located as individual files in <code>input_dir/</code> and labels are defined in <code>fasta-like</code> files in <code>fasta_label_dir/</code> with a <code>.txt</code> ending.</p>
<p>example file <code>file1.fasta</code> in <code>input_dir/</code>:</p>
<pre><code>&gt;fasta_header
ACGTGAGAGAGAGACAGAGATAGACAGAGATTATAA</code></pre>
<p>example label file <code>file1.txt</code> in <code>fasta_label_dir/</code>, where {A,B,C} are the class labels.</p>
<pre><code>&gt;fasta_header
AAAAAAABBBBBBBAAAAAACCCCCCCCCCCCCCCC</code></pre>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1">history &lt;-<span class="st"> </span><span class="kw"><a href="reference/trainNetwork.html">trainNetwork</a></span>(<span class="dt">path =</span> <span class="st">"input_dir"</span>, <span class="dt">labels =</span> <span class="st">"fasta_label_dir/"</span>, <span class="dt">label.vocabulary.size =</span> <span class="dv">3</span>)</a></code></pre></div>
</div>
<div id="datasets" class="section level2">
<h2 class="hasAnchor">
<a href="#datasets" class="anchor"></a>Datasets</h2>
<p>The library comes with three different datasets for testing.</p>
<ul>
<li>The set <code><a href="https://www.rdocumentation.org/packages/utils/topics/data">data(parenthesis)</a></code> contains 100k charakters of the parenthesis synthetic language generated from a very simple counting language with a parenthesis and letter alphabet Σ = {( ) 0 1 2 3 4 }. The language is constrained to match parentheses, and nesting is limited to at most 4 levels deep. Each opening parenthesis increases and each closing parenthesis decreases the nesting level, respectively. Numbers are generated randomly, but are constrained to indicate the nesting level at their position.<br>
</li>
<li>The set <code><a href="https://www.rdocumentation.org/packages/utils/topics/data">data(crispr_full)</a></code> containging all CRISPR loci found in NCBI representative genomes with neighbor nucleotides up and downstream.</li>
<li>The set <code><a href="https://www.rdocumentation.org/packages/utils/topics/data">data(crispr_sample)</a></code> containging a subset of <code>crispr_full</code>
</li>
</ul>
</div>
<div id="license-and-copyright" class="section level2">
<h2 class="hasAnchor">
<a href="#license-and-copyright" class="anchor"></a>License and copyright</h2>
<p>Copyright 2019 Philipp Münch</p>
<p>Source code to deepG is made available under the terms of the <a href="LICENSE.txt">GNU Affero General Public License</a> (AGPL). deepG is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more details.</p>
</div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li>AGPL_3 + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li><a href="authors.html">All authors...</a></li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://travis-ci.org/hiddengenome/deepG"><img src="https://travis-ci.org/hiddengenome/deepG.svg?branch=master" alt="Build Status"></a></li>
<li><a href="https://codecov.io/gh/hiddengenome/deepG"><img src="https://codecov.io/gh/hiddengenome/deepG/branch/master/graph/badge.svg" alt="codecov"></a></li>
<li><a href="https://www.gnu.org/licenses/agpl-3.0"><img src="https://img.shields.io/badge/License-AGPL%20v3-blue.svg" alt="License: AGPL v3"></a></li>
</ul>
</div>
</div>
</div>

      <footer><div class="copyright">
  <p>Developed by .</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
